samza-test/src/main/java/org/apache/samza/test/framework/TestRunner.java

```diff
@@ -267,22 +272,33 @@ public TestRunner addOutputStream(CollectionStream stream) {
     return this;
   }
 
+
   /**
    * Utility to run a test configured using TestRunner
+   *
+   * @param timeout time to wait for the high level application or low level task to finish. This timeout does not include
+   *                input stream initialization time or the assertion time over output streams. This timeout just accounts
+   *                for time that samza job takes run. Samza job won't be invoked with negative or zero timeout
+   * @throws SamzaException if Samza job fails with exception and returns UnsuccessfulFinish as the statuscode
    */
-  public void run() {
+  public void run(Duration timeout) {
     Preconditions.checkState((app == null && taskClass != null) || (app != null && taskClass == null),
```

Comment:
I suggest, we split the validations into simpler ones so that you can precisely convey the error message to the user. The structure right now, forces you to come up with a generic error message for multiple conditions which can be hard to decipher and cryptic to address.

As discussed offline, if you are confident this is an invariant for the `run` method, feel free to remove the condition.

Reply:
I need to see the use cases, if I plan to remove it I will anyways do it in a different PR (cleanup)

Thanks